File: cortex-ts/cortex-web-app/scripts/icon/icon.ts
import * as fs from "fs";
import * as path from "path";
import * as htmlparser2 from "htmlparser2";
import * as css from "css";
import { Buffer } from "buffer";

type resolveFn<InT, OutT extends InT> = (input: InT) => input is OutT;
const makeResolver =
    <InT, OutT extends InT>(condition: boolean): resolveFn<InT, OutT> =>
    (input: InT): input is OutT =>
        condition;

/**
 * Extracts and saves icons from a style element to a specified directory.
 * @param htmlContent - The HTML content containing the style element.
 * @param outputDir - The directory to save the extracted icons.
 */
export async function extractAndSaveIcons(
    htmlContent: string,
    outputDir: string
) {
    // Ensure the output directory exists
    if (!fs.existsSync(outputDir)) {
        fs.mkdirSync(outputDir, { recursive: true });
    }

    // Parse the HTML content to extract the style element content
    let styleContent = "";
    let inStyle = false;

    const parser = new htmlparser2.Parser(
        {
            onopentag(name, attributes) {
                if (name === "style") {
                    inStyle = true;
                }
            },
            ontext(text) {
                if (inStyle) {
                    styleContent += text;
                }
            },
            onclosetag(tagname) {
                if (tagname === "style") {
                    inStyle = false;
                }
            },
        },
        { decodeEntities: true }
    );
    parser.write(htmlContent);
    parser.end();

    // Parse the CSS rules to find background-image URLs
    const cssParsed = css.parse(styleContent);
    const rules = cssParsed.stylesheet?.rules || [];

    rules.forEach((rule) => {
        if (makeResolver<typeof rule, css.Rule>(rule.type === "rule")(rule)) {
            rule.declarations?.forEach((declaration) => {
                if (
                    makeResolver<typeof declaration, css.Declaration>(
                        declaration.type === "declaration"
                    )(declaration) &&
                    declaration.property === "background-image"
                ) {
                    const regex = /url\((data:image\/png;base64,[^)]+)\)/;
                    const match = regex.exec(declaration.value as string);
                    if (match && match[1]) {
                        const base64Data = match[1].replace(
                            /^data:image\/png;base64,/,
                            ""
                        );
                        const buffer = Buffer.from(base64Data, "base64");
                        const iconName =
                            (rule.selectors || [])
                                .join("_")
                                .replace(/[^\w\-]/g, "") + ".png";
                        const outputPath = path.join(outputDir, iconName);
                        fs.writeFileSync(outputPath, buffer);
                        console.log(`Saved icon: ${outputPath}`);
                    }
                }
            });
        }
    });
}

console.log(`Beginning extraction`);

extractAndSaveIcons(
    fs.readFileSync("./src/scripts/icon/iconTest.html", { encoding: "utf-8" }),
    "./src/scripts/icon/test"
).then(() => {
    console.log(`Finished extraction`);
});

File: cortex-ts/cortex-web-app/scripts/icon/icon.go
package main

import (
	"encoding/base64"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

// unescapeCSSClassName handles escape sequences in CSS class names
func unescapeCSSClassName(className string) string {
	re := regexp.MustCompile(`\\([0-9a-fA-F]{2})\s?`)
	return re.ReplaceAllStringFunc(className, func(s string) string {
		code, _ := strconv.ParseInt(s[1:], 16, 32)
		return string(rune(code))
	})
}

// Extract and save icons from a style element to a specified directory
func extractAndSaveIcons(cssFile string, outputDir string) {
	// Read the CSS file
	fileContent, err := ioutil.ReadFile(cssFile)
	if err != nil {
		log.Fatalf("Error reading CSS file: %v", err)
	}

	// Define the regex to extract base64 images and their CSS class names
	re := regexp.MustCompile(`(?m)(\.icon-[^\s]+)\s*\{[^}]*background-image:\s*url\(data:image/png;base64,([^)]*)\)`)

	// Find all matches
	matches := re.FindAllStringSubmatch(string(fileContent), -1)
	if matches == nil {
		log.Fatalf("No matches found in the CSS file")
	}

	for _, match := range matches {
		// Extract the class name and base64 image data
		className := match[1]
		base64Data := match[2]

		// Unescape the class name
		className = unescapeCSSClassName(className)

		// Decode the base64 string
		imgData, err := base64.StdEncoding.DecodeString(base64Data)
		if err != nil {
			log.Printf("Error decoding base64 data for %s: %v", className, err)
			continue
		}

		// Generate the filename from the CSS class name
		filename := strings.TrimPrefix(className, ".") + ".png"
		filepath := filepath.Join(outputDir, filename)

		// Save the image to the specified directory
		err = ioutil.WriteFile(filepath, imgData, 0644)
		if err != nil {
			log.Printf("Error writing image file for %s: %v", className, err)
			continue
		}

		fmt.Printf("Saved %s\n", filepath)
	}
}

func main() {
	// Input CSS file and output directory
	cssFile := "./iconList.css"
	outputDir := "./out"

	// Create the output directory if it doesn't exist
	if err := os.MkdirAll(outputDir, 0755); err != nil {
		log.Fatalf("Error creating output directory: %v", err)
	}

	// Extract and save icons
	extractAndSaveIcons(cssFile, outputDir)
}
